Index: mozilla/nsprpub/pr/src/misc/prnetdb.c	
===================================================================	
--- mozilla/nsprpub/pr/src/misc/prnetdb.c       (revision 158748)	
+++ mozilla/nsprpub/pr/src/misc/prnetdb.c       (working copy)	
@@ -2030,7 +2030,7 @@	
 #endif	
     {	
         PRADDRINFO *res, hints;	
-        PRStatus rv;	
+        int rv;	
 	
         /*	
          * we assume a RFC 2553 compliant getaddrinfo.  this may at some	
